### PR TITLE
fix: improve nullable handling

### PR DESCRIPTION
**Pull request by latonz**

### CODE REVIEW

File: src/Riok.Mapperly.Abstractions/MapperAttribute.cs [hunk 1]
```diff
@@ -28,6 +28,23 @@ public sealed class MapperAttribute : Attribute
     /// </summary>
     public bool EnumMappingIgnoreCase { get; set; }
 
+    /// <summary>
+    /// Specifies the behaviour in the case when the mapper tries to return <c>null</c> in a mapping method with a non-nullable return type.
+    /// If set to <c>true</c> an <see cref="ArgumentNullException"/> is thrown.
+    /// If set to <c>false</c> the mapper tries to return a default value.
+    /// For a <see cref="string"/> this is <see cref="string.Empty"/>,
+    /// for value types <c>default</c>
+    /// and for reference types <c>new()</c> if a parameterless constructor exists or else an <see cref="ArgumentNullException"/> is thrown.
```
Comment (CommonGuy) [inline src/Riok.Mapperly.Abstractions/MapperAttribute.cs:37]: Shouldn't this last case also emit a diagnostic?

File: test/Riok.Mapperly.Tests/Mapping/UserMethodTest.cs [hunk 2]
```diff
@@ -72,6 +72,68 @@ public void WithExistingInstance()
             .Be("target.StringValue = source.StringValue;");
     }
 
+    [Fact]
+    public void NullableToNonNullableShouldThrow()
```
Comment (CommonGuy) [inline test/Riok.Mapperly.Tests/Mapping/UserMethodTest.cs:76]: IMO this should be extracted into a separate class `NullableTest`, as this doesn't have anything to do with user methods.

File: test/Riok.Mapperly.Tests/Mapping/UserMethodTest.cs [hunk 3]
```diff
@@ -72,6 +72,68 @@ public void WithExistingInstance()
...
+            "class B { }");
+
+        TestHelper.GenerateSingleMapperMethodBody(source)
+            .Should()
+            .Be(@"if (source == null)
+        throw new System.ArgumentNullException(nameof(source));
+    var target = new B();
+    return target;".ReplaceLineEndings());
+    }
+
```
Comment (CommonGuy) [inline test/Riok.Mapperly.Tests/Mapping/UserMethodTest.cs:91]: I would also add tests for

- non-nullable to nullable
- nullable to nullable
- cases where the nullable context is disabled only for the source or the target?

**Review by CommonGuy — CHANGES_REQUESTED** [review]

File: test/Riok.Mapperly.Tests/Mapping/UserMethodTest.cs (hunk 3 above)
Comment (latonz) [reply test/Riok.Mapperly.Tests/Mapping/UserMethodTest.cs]: Added.

File: test/Riok.Mapperly.Tests/Mapping/UserMethodTest.cs (hunk 2 above)
Comment (latonz) [reply test/Riok.Mapperly.Tests/Mapping/UserMethodTest.cs]: Moved.

File: src/Riok.Mapperly.Abstractions/MapperAttribute.cs (hunk 1 above)
Comment (latonz) [reply src/Riok.Mapperly.Abstractions/MapperAttribute.cs:37]: Added a diagnostic.

**Review by CommonGuy — APPROVED** [review]

**Pull request merged**